apis/bigqueryanalyticshub/v1alpha1/listing_reference.go

```diff
@@ -0,0 +1,187 @@
+		}
+		return "", fmt.Errorf("reading referenced %s %s: %w", BigQueryAnalyticsHubListingGVK, key, err)
+	}
+	// Get external from status.externalRef. This is the most trustworthy place.
+	actualExternalRef, _, err := unstructured.NestedString(u.Object, "status", "externalRef")
+	if err != nil {
+		return "", fmt.Errorf("reading status.externalRef: %w", err)
+	}
+	if actualExternalRef == "" {
+		return "", fmt.Errorf("BigQueryAnalyticsHubListing is not ready yet")
+	}
+	r.External = actualExternalRef
+	return r.External, nil
+}
+
+// New builds a BigQueryAnalyticsHubListingRef from the Config Connector BigQueryAnalyticsHubListing object.
+func NewBigQueryAnalyticsHubListingRef(ctx context.Context, reader client.Reader, obj *BigQueryAnalyticsHubListing) (*BigQueryAnalyticsHubListingRef, error) {
+	id := &BigQueryAnalyticsHubListingRef{}
+
+	// Get Parent
+	projectRef, err := refsv1beta1.ResolveProject(ctx, reader, obj, obj.Spec.ProjectRef)
+	if err != nil {
+		return nil, err
+	}
+	projectID := projectRef.ProjectID
+	if projectID == "" {
+		return nil, fmt.Errorf("cannot resolve project")
+	}
+	location := obj.Spec.Location
+	id.parent = &BigQueryAnalyticsHubListingParent{ProjectID: projectID, Location: location}
+
+	// Get desired ID
+	resourceID := valueOf(obj.Spec.ResourceID)
+	if resourceID == "" {
+		resourceID = obj.GetName()
+	}
+	if resourceID == "" {
+		return nil, fmt.Errorf("cannot resolve resource ID")
+	}
+
+	// Use approved External
+	externalRef := valueOf(obj.Status.ExternalRef)
+	if externalRef == "" {
+		id.External = asBigQueryAnalyticsHubListingExternal(id.parent, resourceID)
+		return id, nil
+	}
+
+	// Validate desired with actual
+	actualParent, actualResourceID, err := parseBigQueryAnalyticsHubListingExternal(externalRef)
+	if err != nil {
+		return nil, err
+	}
+	if actualParent.ProjectID != projectID {
+		return nil, fmt.Errorf("spec.projectRef changed, expect %s, got %s", actualParent.ProjectID, projectID)
+	}
+	if actualParent.Location != location {
+		return nil, fmt.Errorf("spec.location changed, expect %s, got %s", actualParent.Location, location)
+	}
+	if actualResourceID != resourceID {
+		return nil, fmt.Errorf("cannot reset `metadata.name` or `spec.resourceID` to %s, since it has already assigned to %s",
+			resourceID, actualResourceID)
+	}
+	id.External = externalRef
+	id.parent = &BigQueryAnalyticsHubListingParent{ProjectID: projectID, Location: location}
+	return id, nil
+}
+
+func (r *BigQueryAnalyticsHubListingRef) Parent() (*BigQueryAnalyticsHubListingParent, error) {
+	if r.parent != nil {
+		return r.parent, nil
+	}
+	if r.External != "" {
+		parent, _, err := parseBigQueryAnalyticsHubListingExternal(r.External)
+		if err != nil {
+			return nil, err
+		}
+		return parent, nil
+	}
+	return nil, fmt.Errorf("BigQueryAnalyticsHubListingRef not initialized from `NewBigQueryAnalyticsHubListingRef` or `NormalizedExternal`")
+}
+
+type BigQueryAnalyticsHubListingParent struct {
```

Comment:
Thought: I do feel like 90% of Refs will either have another type of Ref as a parent, or have ProjectIDAndLocationRef as parent

Reply:
From offline -- this struct will probably need to live in a commons package.  I vote to have our tooling generate the changes and I can move it to a common/ shared place once we use the new generated changes.